Yoast/wordpress-seo
# Adapt getProminentWordsForInternalLinking for Japanese

## Context


*

## Summary


This PR can be summarized in the following changelog entry:

* [yoastseo] Adapts `getProminentWordsForInternalLinking` research for Japanese.
* [yoastseo] Adds `customGetStemmer` and `determineStem` helper for Japanese.

## Relevant technical choices:

* For Japanese, we don't retrieve abbreviation. As such, we assign it to an empty array.
* A helper to determine the canonical stem of a word in Japanese is created. The canonical stem is the shortest word/form from the array of forms that were created in `createWordForms`.
* A custom helper (`customGetStemmer`) is also created where it returns the helper function that returns canonical stem. For Japanese, this helper will be used for Prominent words functionality instead of the regular `getStemmer` helper.
* In Japanese, we only want to return suggestions (and spend time calculating prominent words) if the text is at least 200 characters instead of 100 words. The number 200 is picked as in Japanese for assessment that uses text length, we decided to have 1:2 word-character ratio (see [the epic description](https://yoast.atlassian.net/browse/PRODUCT-5)).

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

* Build Free plugin
* Run `composer require yoast/wordpress-seo:dev-LINGO-1095-adapt-getProminentWordsForInternalLinking-for-japanese@dev` in `wordpress-seo-premium`
* Build Premium plugin
* Enable the `JAPANESE_SUPPORT` feature by adding `define( 'YOAST_SEO_JAPANESE_SUPPORT', true );` in `basic-wordpress-config.php` file
* Set your site to Japanese` 日本語`
* Create two posts with Japanese characters (tip: use the text generated by [this generator](https://generator.lorem-ipsum.info/_japanese))
* Make sure that the two posts are identical, including the title, and that in the text the same string is repeated at least twice per paragraph (e.g., `発少わぽぼン`)
* Open one of the posts
* Expand the Internal linking suggestion tab in the sidebar
* Confirm that the other post that you created is suggested in the Internal linking suggestion. It should look something like this:
<img width="287" alt="Screenshot 2021-11-16 at 10 45 15" src="https://user-images.githubusercontent.com/48715883/141961794-45f57df1-c30b-493e-a428-d55e4facd477.png">

* Disable the `JAPANESE_SUPPORT` by adding this  `define( 'YOAST_SEO_JAPANESE_SUPPORT', false );` in `basic-wordpress-config.php` file
* Reload your post or Close and reopen it
* Confirm that in Internal linking suggestion tab, no other post is suggested to link. It should look like this:
<img width="290" alt="Screenshot 2021-11-16 at 09 26 23" src="https://user-images.githubusercontent.com/48715883/141962152-5f30ec40-0fa2-4262-8e99-e21a8f744cf3.png">

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* Install and activate Free and Premium plugin
* Enable the `JAPANESE_SUPPORT` feature by adding `define( 'YOAST_SEO_JAPANESE_SUPPORT', true );` in your `wp-config.php` file for your setup
* Set your site to Japanese` 日本語`
* Create two posts with Japanese characters (tip: use the text generated by [this generator](https://generator.lorem-ipsum.info/_japanese))
* Make sure that the two posts are identical, including the title, and that in the text the same string is repeated at least twice per paragraph (e.g., `発少わぽぼン`)
* Open one of the posts
* Expand the Internal linking suggestion tab in the sidebar
* Confirm that the other post that you created is suggested in the Internal linking suggestion. It should look something like this:
<img width="287" alt="Screenshot 2021-11-16 at 10 45 15" src="https://user-images.githubusercontent.com/48715883/141961794-45f57df1-c30b-493e-a428-d55e4facd477.png">

* Disable the `JAPANESE_SUPPORT` by adding this  `define( 'YOAST_SEO_JAPANESE_SUPPORT', false );` in your `wp-config.php` file for your setup
* Reload your post or Close and reopen it
* Confirm that in Internal linking suggestion tab, no other post is suggested to link. It should look something like this:
<img width="290" alt="Screenshot 2021-11-16 at 09 26 23" src="https://user-images.githubusercontent.com/48715883/141962152-5f30ec40-0fa2-4262-8e99-e21a8f744cf3.png">

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Other environments


## Documentation


## Quality assurance


Fixes https://yoast.atlassian.net/browse/LINGO-1095
